### PR TITLE
Add offers - complete changeset

### PR DESCRIPTION
**Pull request by jackcarlisle**
# ready for review

- [x] adds functionality to create offers

#12

### CODE REVIEW

File: web/controllers/helpers.ex [hunk 1]
```diff
@@ -2,6 +2,21 @@ defmodule Karma.Controllers.Helpers do
...
+  #     "equipment_rental_fee_per_week",
+  #     "vehicle_allowance_per_week",
+  #     "fee_per_day_inc_holiday"
+  #   ]
+  #   for integer_string <- integer_strings do
+  #     Map.put(offer_map, integer_string, String.to_integer(offer_map[integer_string]))
+  #   end
+  #   offer_map
+  # end
+
```
Comment (JMurphyWeb) [inline web/controllers/helpers.ex]: can we delete these?

Comment (jackcarlisle) [inline web/controllers/helpers.ex]: yep! done 👍

File: web/controllers/offer_controller.ex [hunk 2]
```diff
@@ -43,7 +43,52 @@ defmodule Karma.OfferController do
...
+    "seventh_day_fee_multiplier" => seventh_day_fee_multiplier} = offer_params
+
+    sixth_day_fee_multiplier = String.to_float(sixth_day_fee_multiplier)
+    seventh_day_fee_multiplier = String.to_float(seventh_day_fee_multiplier)
+
+    working_week =
+      case working_week do
+        "" -> ""
+        _ww -> String.to_float(working_week)
+      end
```
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: are working_week & fee_per_day_inc_holiday are both required fields?

Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: It looks like they are. I think we can run the user's submitted data through a validation changeset before doing any calculation, we can reduce a load of checks, both in this file and the helpers?

Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: yep

File: web/controllers/offer_controller.ex [hunk 3]
```diff
@@ -43,7 +43,52 @@ defmodule Karma.OfferController do
...
+    "fee_per_day_inc_holiday" => fee_per_day_inc_holiday,
+    "working_week" => working_week,
+    "department" => department,
+    "job_title" => job_title,
+    "sixth_day_fee_multiplier" => sixth_day_fee_multiplier,
+    "seventh_day_fee_multiplier" => seventh_day_fee_multiplier} = offer_params
+
+    sixth_day_fee_multiplier = String.to_float(sixth_day_fee_multiplier)
+    seventh_day_fee_multiplier = String.to_float(seventh_day_fee_multiplier)
+
```
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: I think these can be done automatically by our validation changeset function.

Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: would the changeset not just check to see if it's the right type?

File: web/controllers/offer_controller.ex (hunk 2 above)
Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: yeah I agree! Definitely needs a refactor

File: web/controllers/offer_controller.ex [hunk 4]
```diff
@@ -43,7 +43,52 @@ defmodule Karma.OfferController do
...
+    holiday_pay_per_day = calc_holiday_pay_per_day(fee_per_day_inc_holiday, fee_per_day_exc_holiday)
+    fee_per_week_inc_holiday = calc_fee_per_week_inc_holiday(fee_per_day_inc_holiday, working_week)
+    fee_per_week_exc_holiday = calc_fee_per_week_exc_holiday(fee_per_week_inc_holiday, project.holiday_rate)
+    holiday_pay_per_week = calc_holiday_pay_per_week(fee_per_week_inc_holiday, fee_per_week_exc_holiday)
+    contract_type = determine_contract_type(department, job_title)
+    sixth_day_fee_inc_holiday = calc_day_fee_inc_holidays(fee_per_day_inc_holiday, sixth_day_fee_multiplier)
+    sixth_day_fee_exc_holiday = calc_day_fee_exc_holidays(fee_per_day_exc_holiday, sixth_day_fee_multiplier)
+    seventh_day_fee_inc_holiday = calc_day_fee_inc_holidays(fee_per_day_inc_holiday, seventh_day_fee_multiplier)
+    seventh_day_fee_exc_holiday = calc_day_fee_exc_holidays(fee_per_day_exc_holiday, seventh_day_fee_multiplier)
+
```
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: We can wrap all of these up into a separate function which takes our (already validated) map and returns another map. 

```elixir
calculated_values = get_calculated_values(validated_offer_values, project)
```
so calculated_values looks something like this:
```elixir
%{fee_per_day_exc_holiday: 22
 , holiday_pay_per_day: 343
 , ...
 }
```

Then I think we can replace all of the following lines with:
```elixir
Map.merge(validated_offer_values, calculated_values)
```

I may have missed something

File: web/controllers/offer_controller.ex (hunk 2 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: @jackcarlisle the changeset that is there should be fine, I think we can just move it up a bit and then check if it is valid or not before moving onto the calculations?

File: web/controllers/offer_controller.ex (hunk 3 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: not sure. ill check now

File: web/controllers/offer_controller.ex (hunk 4 above)
Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: I tried to implement this way first time around but kept running into issues. I want to refactor everything next week, just wanted to get the flow working for now

Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: https://github.com/karmaradio/karma/issues/88

Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: We can put the `get_calculated_values` function in a separate module along with the actual calculator functions that you have built.

Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: yeah that's what I was thinking too

File: web/controllers/offer_controller.ex (hunk 3 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]:
```elixir
iex(2)> result = User.test(%User{}, %{string: "1"})
#Ecto.Changeset<action: nil, changes: %{string: 1.0}, errors: [],
 data: #Karma.User<>, valid?: true>
```

I just added a field into the user model with name :string and type :float.  Then made a test changeset which just cast that :string field. Seems to convert it into a float.

File: web/controllers/offer_controller.ex (hunk 4 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: If its working now, it should be a pretty straightforward refactor. Would be better to do it as part of this PR cos that create function is a bit meaty. I don't mind doing the refactor, it shouldn't take long should it?

Comment (jackcarlisle) [inline web/controllers/offer_controller.ex]: yeah shouldn't take too long! If you're happy to then that's cool go for it!

File: web/controllers/offer_controller.ex (hunk 3 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: done in #89 

changeset.changes holds the parsed values so not perfect, but good for now.

File: web/controllers/offer_controller.ex (hunk 2 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: done in #89

File: web/controllers/offer_controller.ex (hunk 4 above)
Comment (JMurphyWeb) [inline web/controllers/offer_controller.ex]: Done in #89

**Comment by JMurphyWeb** [comment]
It took longer than I thought!!

Take a look over #89 and see if it all looks ok, I may have missed something

**Pull request merged**